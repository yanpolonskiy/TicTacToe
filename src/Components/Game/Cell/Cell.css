.cell {
  height: 50px;
  flex: 0 0 50px;
  background-color: #dddddd;
  border: 1px solid black;
  padding: 0;
  table-layout: fixed;
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
}

.cell:hover {
  background-color: #e7e7e7;
}
