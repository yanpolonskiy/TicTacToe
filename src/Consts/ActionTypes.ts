/**
 * Енам игровых экшенов
 */
export enum GAME_ACTION {
  MAKE_MOVE = `MAKE_MOVE`,
}

/**
 * Вспомогательный тип для тайпгварда типа экшена и его payload.
 */
type Action<K = GAME_ACTION, V = void> = V extends void
  ? { type: K }
  : { type: K } & V;

/**
 * Тип координат.
 */
export type TCoordinates = [number, number];

/**
 * Интерфейс хода.
 * Сделано в виде обьекта для удобного добавления дополнительной информации в ход.
 */
export interface IMove {
  cell: TCoordinates;
}

/**
 * Тайпгвард игровых экшенов
 */
export type TGameActions = Action<GAME_ACTION.MAKE_MOVE, { payload: IMove }>;

/**
 * Маппинг для класса игровых экшенов.
 */
export interface IGameActions {
  [prop: string]: (props: any) => TGameActions;
}
