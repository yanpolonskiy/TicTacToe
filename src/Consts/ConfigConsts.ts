/**
 * Константа ширины клетки, для расчета отрисовки игрового поля.
 */
export const CellWidth = 50;

/**
 * Константа высоты клетки, для расчета отрисовки игрового поля.
 */
export const CellHeight = 50;
