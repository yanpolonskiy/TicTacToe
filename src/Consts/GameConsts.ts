/**
 * Енам ходов.
 * X - Значение хода крестик.
 * O - Значение хода нолик.
 */
export enum EValue {
  X = "X",
  O = "O",
}
