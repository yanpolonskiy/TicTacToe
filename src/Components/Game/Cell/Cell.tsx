import React from "react";
import { EValue } from "Consts/GameConsts";
import "./Cell.css";
import {TCoordinates} from 'Consts/ActionTypes';

/**
 * Интерфейс клетки игрового поля.
 * @prop {EValue} [value] Значение в клетке.
 * @prop {TCoordinates} coordinates Координаты клетки.
 * @prop {Function} onClick Функция коллбек клика по клетке.
 */
interface IProps {
  value?: EValue;
  coordinates: TCoordinates;
  onClick: (coordinate: TCoordinates) => void;
}

export class Cell extends React.Component<IProps> {
  /**
   * Обработчик клика по клетке.
   */
  handleClick = () => {
    this.props.onClick(this.props.coordinates);
  };

  render() {
    const { value } = this.props;
    const valueClassName = `value ${
      value ? (value === EValue.X ? "x-value" : "o-value") : ""
    }`;
    return (
      <div className="cell" onClick={this.handleClick}>
        <span className={valueClassName}>{value}</span>
      </div>
    );
  }
}
