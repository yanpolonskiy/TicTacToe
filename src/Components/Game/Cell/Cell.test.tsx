import React from 'react';
import {Cell} from './Cell';
import {EValue} from 'Consts/GameConsts';
import renderer from 'react-test-renderer';
import {TCoordinates} from '../../../Consts/ActionTypes';

/**
 * Тест отрисовки клетки Cell.
 */
test("renders cell", () => {
    const tree = renderer
    .create(<Cell coordinates={[1,2]} value={EValue.X} onClick={(coordinates: TCoordinates) => coordinates}/>)
    .toJSON();
    expect(tree).toMatchSnapshot();
});
