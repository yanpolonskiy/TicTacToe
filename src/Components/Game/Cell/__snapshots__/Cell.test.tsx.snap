// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders cell 1`] = `
<div
  className="cell"
  onClick={[Function]}
>
  <span
    className="value x-value"
  >
    X
  </span>
</div>
`;
